### ISSUE TITLE
data contamination with Shuffle=True in train_test_split

### ISSUE THREAD

**Issue by ThomasGoud** [nateemma/strategies#8]
Hello,

Thank you for your incredible repository very interesting!
When analysing your code, especially binanceus/PCA.py, I'm wondering why the train/test dataframe for model training is shuffle:
       ` df_train, df_test, res_train, res_test = train_test_split(df, labels, train_size=0.8, random_state=27, shuffle=True)`

Isn't there a risk a contaminating the model with future data (looking into the future)?

Thanks,
Thomas

**Comment by ThomasGoud** [comment]
Hello Phil,

Thanks for your reply, after inspecting your code and rereading your explanation it is now clear how it is working.

The only blackpoint I see, if I understood well, is that the backtesting will occur on points where the classifier has been trained. 

Do you consider that these codes can be run in trading mode (dry_run)? Because it seems that populate_indicators() method will be called every 5 second and the PCA / Classifier training will be made over and over. 
Would you advice to add a time check to restart the PCA / CLF training every 1h for example? 
Finally, wouldn't it help to avoid overfitting to have only one PCA/CLF over every coins instead of one per coin? 

Thanks in advance,

Thomas

**Comment by nateemma** [comment]
Again, those are very good questions, some answers below:

*The only blackpoint I see, if I understood well, is that the backtesting
will occur on points where the classifier has been trained.*

Yes, but that's the only data I have, so I can't avoid it. For the PCA
algorithms, I take a random sampling of available data for fitting, so it
doesn't really match the data that would be seen when assessing buy/sell
criteria (but it is not the best situation). This is actually why I started
looking at alternate types of strategy such as neural networks and anomaly
detection.

For the neural network strategies, the training doesn't happen on all of
the data, some of it is reserved for testing - you can also train on older
data, but test on newer data, which the training run didn't see (I started
doing this recently). Also, neural networks don't really 'remember' the
data, they just set the model parameters to values that produce the best
results (lowest loss) when comparing predicted values to actual values (the
buy/sell signals created by looking ahead). Additionally, see the answer to
your last question.

*Do you consider that these codes can be run in trading mode (dry_run)?
Because it seems that populate_indicators() method will be called every 5
second and the PCA / Classifier training will be made over and over.*

Training for neural network models only happens in backtest mode. For the
PCA strategies, I cannot avoid this because the underlying classifiers do
not support cumulative fitting, i.e. I have to refit every time otherwise I
wouldn't have a valid model. However, these algorithms are quite fast, and
I have not had any issues running in trading modes.

*Would you advice to add a time check to restart the PCA / CLF training
every 1h for example?*

I have not seen a need to do this. However, there is already some
randomness built in, where refitting does not happen on every call (a
random number is generated after fitting, and re-fitting does not occur
again until that many calls have happened).


*Finally, wouldn't it help to avoid overfitting to have only one PCA/CLF
over every coins instead of one per coin?*

Unfortunately, this is not possible for the PCA (or most of the Anomaly
Detection) algorithms, because the underlying classifiers cannot be
cumulatively fitted - so I cannot combine data from one pair with data from
another.

For the neural network strategies (NNBC_*, NNPredict_*) this option does
exist (just set model_per_pair to False, which is the default)

Hope that helps

Thanks,

Phil
